keps/sig-instrumentation/1753-logs-sanitization/README.md

```diff
@@ -0,0 +1,243 @@
+
+When it comes to standard go lang types and third party libraries used in Kubernetes which cannot be changed we propose listing them in one place with information about the type of the sensitive data which they contain. 
+
+We also propose to implement a small library which could use the above information to verify if any of the provided values contain reference to sensitive data.
+
+Finally we propose to integrate this library with the klog logging library used by Kubernetes in a way that when enabled the log entries which contain information marked as sensitive will be redacted from the logs.
+
+### Risks and Mitigations
+
+#### Performance overhead
+
+Inspection of log parameters can be time consuming and it can impose significant performance overhead on each log library invocation.
+
+This risk can be mitigated by:
+- running inspection only on log entry parameters which are going to be actually logged - running inspection after log v-level has been evaluated.
+- introducing a dedicated go lang interface with a single method returning information about types of sensitive data which given value contains like:
+
+  ```go
+  type DatapolicyDetector interface {
+	  ContainsDatapolicyValues() []string
+  }
+  ```
+
+- implementations of this interface could be auto generated using a dedicated code generator similar to deep-copy generator or manually implemented when needed,
+caching negative inspection results for parameter types which does not have any references to types which may contain sensitive data.
+
+Which of those methods will be used and to what extent will be decided after running performance tests.
+
+
+## Design Details
+
+### Source code tags
+
+We propose to mark all struct fields which may contain sensitive data with a new datapolicy tag which as a value will accept a list of predefined types of data.
+
+For now we propose following types of data to be available as values:
+- password
+- token
+- security-key
+
+Example of using datapolicy tag:
+
+```go
+type ConfigMap struct {
+  Data map[string]string `json:"data,omitempty" datapolicy:”password,token,security-key”`
+}
+```
+
+For external types which are not part of the Kubernetes source code such as go lang standard libraries or third party vendored libraries for which we cannot change source code we will define a global function which will map those into relevant datapolicy tag:
+
+```go
+func GlobalDatapolicyMapping(val interface{}) []string
+```
+
+### datapolicy verification library
+
+datapolicy verification library will implement verification if the provided value does not contain any sensitive data identified by the datapolicy tag. Verification will be performed using reflection. Recursion will stop on pointers as values for those are usually not logged. Validation will depend only on the datapol tags values. Values of individual fields of primitive types like string will not be analyzed other than checking if a given field is not empty.
```

Comment:
"datapolicy verification library will implement verification if the provided value does not contain any sensitive data identified by the datapolicy tag." doesn't make sense to me. 

Don't you have to perform verification in order to figure out if it may contain sensitive data to begin with? Am I misunderstanding this?

Reply:
Would you mind explicitly drawing out the difference between verification and validation (if there is one)?

Reply:
There is no difference between verification and validation in this context. I have updated a description a bit to make it clearer and aligned it to use verification term.